### PR TITLE
Add Microsoft Hypervisor as a supported hypervisor

### PR DESCRIPTION
**Pull request by russell-islam**

### CODE REVIEW

File: hypervisor/src/mshv/mod.rs [hunk 1]
```diff
@@ -0,0 +1,10 @@
+// SPDX-License-Identifier: Apache-2.0 OR BSD-3-Clause
+//
+// Copyright © 2020, Microsoft Corporation
+//
+
+#![allow(dead_code)]
+#![allow(unused_imports)]
+#![allow(unused_variables)]
+#![allow(unused_macros)]
+#![allow(non_upper_case_globals)]
```
Comment (liuw) [inline hypervisor/src/mshv/mod.rs:10]: These were added to help focus on development. At the time I implemented more things than necessary because I wanted them to be available whenever I needed them.

At least the first four should be dropped when it comes to upstreaming and we should fix whatever warnings that show up. I don't think there are that many.

File: hypervisor/src/mshv/x86_64/mod.rs [hunk 2]
```diff
@@ -0,0 +1,154 @@
...
+        msr!(msr_index::MSR_IA32_SYSENTER_CS),
+        msr!(msr_index::MSR_IA32_SYSENTER_ESP),
+        msr!(msr_index::MSR_IA32_SYSENTER_EIP),
+        msr!(msr_index::MSR_STAR),
+        msr!(msr_index::MSR_CSTAR),
+        msr!(msr_index::MSR_LSTAR),
+        msr!(msr_index::MSR_KERNEL_GS_BASE),
+        msr!(msr_index::MSR_SYSCALL_MASK),
+        msr!(msr_index::MSR_IA32_TSC),
+        // TODO: feature guard kvm and move this out of kvm
```
Comment (liuw) [inline hypervisor/src/mshv/x86_64/mod.rs:146]: This comment seems to be stale.

File: hypervisor/src/lib.rs [hunk 3]
```diff
@@ -62,5 +62,5 @@ pub fn new() -> std::result::Result<Arc<dyn Hypervisor>, HypervisorError> {
 
     Ok(Arc::new(hv))
 }
-#[cfg(all(feature = "mshv", target_arch = "x86_64"))]
+
```
Comment (liuw) [inline hypervisor/src/lib.rs]: Unrelated change to this particular patch.

(Not sure how this will show up in GitHub UI)

Comment (liuw) [reply hypervisor/src/lib.rs]: Also, there is an extraneous blank line between the two SoB lines.

File: hypervisor/src/mshv/mod.rs [hunk 4]
```diff
@@ -13,6 +13,18 @@
 #[cfg(target_arch = "x86_64")]
 pub mod x86_64;
 
+use mshv_bindings::Msrs as MsrEntries;
+pub use mshv_bindings::*;
+use mshv_ioctls::{set_registers_64, Mshv, VcpuFd, VmFd};
+
+/// Vcpu struct for Hyper-V
```
Comment (liuw) [inline hypervisor/src/mshv/mod.rs:384]: Hyper-V should be Microsoft Hypervisor. Please also fix other occurrences in code and commit message.

Comment (liuw) [reply hypervisor/src/mshv/mod.rs:384]: And again, there is an extraneous line between two SoB lines.

File: hypervisor/src/vm.rs [hunk 5]
```diff
@@ -220,6 +229,18 @@ pub trait Vm: Send + Sync {
     fn check_extension(&self, c: Cap) -> bool;
     /// Create a device that is used for passthrough
     fn create_passthrough_device(&self) -> Result<Arc<dyn Device>>;
+    /// XXX temp workaround for Hyper-V, never used on KVM
```
Comment (liuw) [inline hypervisor/src/vm.rs:232]: Please turn this `XXX` into a proper comment.

File: hypervisor/src/mshv/mod.rs [hunk 6]
```diff
@@ -586,6 +586,117 @@ impl cpu::Vcpu for MshvVcpu {
     }
 }
 
+struct MshvEmulatorContext<'a> {
```
Comment (liuw) [inline hypervisor/src/mshv/mod.rs:776]: Sameul and I wrote this patch so you should drop Nuno and Praveen's tags.

**Review by liuw — COMMENTED** [review]
Thanks for doing this, Muminul.

I don't have much to say since I've written a big portion of this series, so of course I approved the general direction. ;-)

I commented on individual patches. Not sure how they will show up in GitHub's UI, but if you can, navigate to individual patches may make my comments make more sense to you.

File: hypervisor/src/mshv/mod.rs [hunk 7]
```diff
@@ -52,6 +51,205 @@ pub struct HvState {
...
+        return true;
+    }
+
+    false
+}
+
+fn get_delivery_mode(message_data: u32) -> u8 {
+    ((message_data & 0x700) >> 8) as u8
+}
+
```
Comment (sameo) [inline hypervisor/src/mshv/mod.rs:204]: Maybe it would be cleaner to make those `MshvIrqRoutingMsi` methods instead of having them floating around the code?

File: hypervisor/src/mshv/mod.rs [hunk 8]
```diff
@@ -52,6 +51,205 @@ pub struct HvState {
...
+// True => High active
+// False => Low active
+fn get_level(message_data: u32) -> bool {
+    if (message_data >> 14) & 0x1 == 0x1 {
+        return true;
+    }
+
+    false
+}
+
```
Comment (sameo) [inline hypervisor/src/mshv/mod.rs:215]: I think this one is not used.

File: hypervisor/src/mshv/mod.rs [hunk 9]
```diff
@@ -526,7 +526,154 @@ impl cpu::Vcpu for MshvVcpu {
...
+                        if let Some(vmmops) = &self.vmmops {
+                            vmmops
+                                .pio_read(port.into(), &mut data[0..len])
+                                .map_err(|e| cpu::HypervisorCpuError::RunVcpu(e.into()))?;
+                        }
+                        // debug!("data {:x?}", &data[0..len]);
+                        let v = data[0] as u32
+                            | (data[1] as u32) << 8
+                            | (data[2] as u32) << 16
+                            | (data[3] as u32) << 24;
```
Comment (sameo) [inline hypervisor/src/mshv/mod.rs:582]: This is `let v = u32::from_le_bytes(data);`

File: hypervisor/src/mshv/mod.rs [hunk 10]
```diff
@@ -526,7 +526,154 @@ impl cpu::Vcpu for MshvVcpu {
...
+                    //     "port {:x?} insn byte count {:?} len {:?} write {:?}",
+                    //     port, info.instruction_byte_count, len, is_write
+                    // );
+
+                    if is_write {
+                        // debug!("data {:x?}", info.rax);
+                        data[0] = info.rax as u8;
+                        data[1] = (info.rax >> 8) as u8;
+                        data[2] = (info.rax >> 16) as u8;
+                        data[3] = (info.rax >> 24) as u8;
```
Comment (sameo) [inline hypervisor/src/mshv/mod.rs:566]: `let data = (info.rax as u32).to_le_bytes()`

File: hypervisor/src/mshv/mod.rs [hunk 11]
```diff
@@ -0,0 +1,10 @@
+// SPDX-License-Identifier: Apache-2.0 OR BSD-3-Clause
```
[the remark this answers is not in the document]
Comment (russell-islam) [reply hypervisor/src/mshv/mod.rs:1]: Not really, we followed the same as before. If you think this is something we need to revisit please sync up with Guru.

File: hypervisor/src/mshv/mod.rs [hunk 12]
```diff
@@ -52,6 +51,205 @@ pub struct HvState {
...
+            self.irqfd.as_raw_fd(),
+            epoll::Events::EPOLLIN,
+            u64::from(IRQFD_EVENT),
+        )
+        .unwrap_or_else(|err| {
+            info!(
+                "IrqfdCtrlEpollHandler: failed to register listener: {:?}",
+                err
+            );
+        });
```
[the remark this answers is not in the document]
Comment (russell-islam) [reply hypervisor/src/mshv/mod.rs:109]: I would let @liuw  to comment.

File: hypervisor/src/mshv/mod.rs (hunk 7 above)
Comment (russell-islam) [reply hypervisor/src/mshv/mod.rs:204]: This implementation mainly contributed by @liuw >I would let him comment.

File: hypervisor/src/mshv/mod.rs [hunk 13]
```diff
@@ -52,6 +51,205 @@ pub struct HvState {
...
+        4 => Some(hv_interrupt_type_HV_X64_INTERRUPT_TYPE_NMI),
+        5 => Some(hv_interrupt_type_HV_X64_INTERRUPT_TYPE_INIT),
+        7 => Some(hv_interrupt_type_HV_X64_INTERRUPT_TYPE_EXTINT),
+        _ => None,
+    }
+}
+
+// See Intel SDM vol3 10.11.1
+// We assume APIC ID and Hyper-V Vcpu ID are the same value
+// This holds true for HvLite
```
[the remark this answers is not in the document]
Comment (liuw) [reply hypervisor/src/mshv/mod.rs:167]: HvLite is a project that is internal to Microsoft. This comment should be dropped.

File: hypervisor/src/mshv/mod.rs (hunk 7 above)
Comment (liuw) [reply hypervisor/src/mshv/mod.rs:204]: Yes. Making them methods of `MshvIrqRoutingMsi` is fine. They are not tied to MSHV in anyway, but making them methods won't hurt either.

File: hypervisor/src/kvm/mod.rs [hunk 14]
```diff
@@ -159,7 +159,7 @@ impl vm::Vm for KvmVm {
     ///
     /// Registers an event that will, when signaled, trigger the `gsi` IRQ.
     ///
-    fn register_irqfd(&self, fd: &EventFd, gsi: u32) -> vm::Result<()> {
+    fn register_irqfd(&self, fd: &EventFd, gsi: u32, _vm: Arc<dyn vm::Vm>) -> vm::Result<()> {
```
[the remark this answers is not in the document]
Comment (russell-islam) [reply hypervisor/src/kvm/mod.rs:162]: I do agree with you @sameo. Let get thought from @liuw

File: hypervisor/src/mshv/mod.rs (hunk 11 above)
[the remark this answers is not in the document]
Comment (russell-islam) [reply hypervisor/src/mshv/mod.rs:1]: @gbalane Can also comment here.

File: hypervisor/src/mshv/mod.rs (hunk 12 above)
[the remark this answers is not in the document]
Comment (liuw) [reply hypervisor/src/mshv/mod.rs:109]: No. I don't think it makes `run_ctrl()` exit. The thread will be just hanging forever because it will block on those eventfds.

I was not sure at the time there would be a better way to handle this. I didn't want to exit the thread because I was worried that it might cause failure when `unregister_irqfd` wrote to the `kill_fd`. Now that I think about this, it is probably okay to let the thread exit because `kill_fd` will hang on for as long as needed.

Note that the emulation is just a stop-gap. It is going to be superseded by our work to implement irqfd and ioeventfd inside Linux kernel. We already wrote some patches and they will be posted once ready.

File: hypervisor/src/kvm/mod.rs (hunk 14 above)
[the remark this answers is not in the document]
Comment (liuw) [reply hypervisor/src/kvm/mod.rs:162]: I agree with Samuel. The code could use a bit of restructuring.

File: hypervisor/src/mshv/mod.rs (hunk 12 above)
[the remark this answers is not in the document]
Comment (sameo) [reply hypervisor/src/mshv/mod.rs:150]: > Note that the emulation is just a stop-gap. It is going to be superseded by our work to implement irqfd and ioeventfd inside Linux kernel. We already wrote some patches and they will be posted once ready.

Then let's keep it as is for now.

File: hypervisor/src/mshv/mod.rs (hunk 1 above)
Comment (russell-islam) [reply hypervisor/src/mshv/mod.rs:10]: I was thinking to do it in a future PR if you do agree.

Comment (liuw) [reply hypervisor/src/mshv/mod.rs:10]: Fine by me.

**Comment by liuw** [comment]
@sameo @rbradford 

The clippy warning is not straight-forward to fix, because `--all-features` enables both `mshv` and `kvm`. Cargo doesn't have a way to specify mutually exclusive features.

I've written a patch. See https://github.com/liuw/cloud-hypervisor/commit/6df3afda232f2a2049953b16f895fae7e44e79de .

I would like some input from you both on whether that's acceptable.

**Comment by sameo** [comment]
> @sameo @rbradford
> 
> The clippy warning is not straight-forward to fix, because `--all-features` enables both `mshv` and `kvm`. Cargo doesn't have a way to specify mutually exclusive features.
> 
> I've written a patch. See [liuw@6df3afd](https://github.com/liuw/cloud-hypervisor/commit/6df3afda232f2a2049953b16f895fae7e44e79de) .
> 
> I would like some input from you both on whether that's acceptable.

What about that one https://github.com/sameo/cloud-hypervisor-1/commit/3cd4916ce1283d7bdd410027d2ba1865b199c4b6 ?

**Comment by liuw** [comment]
> 
> 
> > @sameo @rbradford
> > The clippy warning is not straight-forward to fix, because `--all-features` enables both `mshv` and `kvm`. Cargo doesn't have a way to specify mutually exclusive features.
> > I've written a patch. See [liuw@6df3afd](https://github.com/liuw/cloud-hypervisor/commit/6df3afda232f2a2049953b16f895fae7e44e79de) .
> > I would like some input from you both on whether that's acceptable.
> 
> What about that one [sameo@3cd4916](https://github.com/sameo/cloud-hypervisor-1/commit/3cd4916ce1283d7bdd410027d2ba1865b199c4b6) ?

This looks better. Let's go with yours.

**Comment by russell-islam** [comment]
@sameo @liuw The clippy warnigs have been resolved now. Only error we are geeting is as follow 
`thread 'rustc' panicked at 'called `Option::unwrap()` on a `None` value', compiler/rustc_middle/src/ty/query/mod.rs:235:5`
I this this is a known issue upstream.

Also I have addressed all of your comments. If I missed anything that is not intentional. Please let me know if you still have some comments. I will address those ASAP and try to get this PR merged at my earliest.

File: hypervisor/src/cpu.rs [hunk 15]
```diff
@@ -272,10 +272,12 @@ pub trait Vcpu: Send + Sync {
     /// Setup the model-specific registers (MSR) for this vCPU.
     ///
     fn set_msrs(&self, msrs: &MsrEntries) -> Result<usize>;
+    #[cfg(feature = "kvm")]
```
[the remark this answers is not in the document]
Comment (liuw) [reply hypervisor/src/cpu.rs:294]: There is a word "bas" which should perhaps be deleted?

File: hypervisor/src/mshv/x86_64/mod.rs [hunk 16]
```diff
@@ -0,0 +1,154 @@
...
+pub struct MpState {}
+
+#[derive(Eq, PartialEq, Hash, Clone, Debug, Copy)]
+pub enum IoEventAddress {
+    /// Representation of an programmable I/O address.
+    Pio(u64),
+    /// Representation of an memory mapped I/O address.
+    Mmio(u64),
+}
+macro_rules! msr {
```
[the remark this answers is not in the document]
Comment (russell-islam) [reply hypervisor/src/mshv/x86_64/mod.rs:51]: Will do in a future PR.

File: hypervisor/src/mshv/mod.rs (hunk 11 above)
[the remark this answers is not in the document]
Comment (russell-islam) [reply hypervisor/src/mshv/mod.rs:1]: We can revisit this later if needed.

File: hypervisor/src/mshv/mod.rs [hunk 17]
```diff
@@ -464,12 +478,13 @@ impl vm::Vm for MshvVm {
     /// Get the Vm state. Return VM specific data
     ///
     fn state(&self) -> vm::Result<VmState> {
-        unimplemented!();
+        Ok(*self.hv_state.read().unwrap())
```
[the remark this answers is not in the document]
Comment (russell-islam) [reply hypervisor/src/mshv/mod.rs:1053]: Yes. This is needed for restore and snapshot. This is VM specific data.  hv_state is named after the hv interface documentationjs.

File: hypervisor/src/mshv/mod.rs (hunk 11 above)
[the remark this answers is not in the document]
Comment (gbalane) [reply hypervisor/src/mshv/mod.rs:1]: > @gbalane Can also comment here.

We are fine with Dual Licensing and it is inline with CLH based on the previous discussion.

**Comment by sameo** [comment]
We worked with @russell-islam on the following, additional changes:

- Clean the commit messages up in order for them to adhere to https://www.kernel.org/doc/Documentation/process/submitting-patches.rst wrt to SOBs and Co-developed-by tags
- Squash 35bf055e27f7867654a506fa0c46302646d434a2 with 4bb31daa18e5d7181c7d4e1388e101b5536db463
- Include missing Cargo.lock changes

**Review by sameo — APPROVED** [review]

**Comment by russell-islam** [comment]
> We worked with @russell-islam on the following, additional changes:
> 
>     * Clean the commit messages up in order for them to adhere to https://www.kernel.org/doc/Documentation/process/submitting-patches.rst wrt to SOBs and Co-developed-by tags
> 
>     * Squash [35bf055](https://github.com/cloud-hypervisor/cloud-hypervisor/commit/35bf055e27f7867654a506fa0c46302646d434a2) with [4bb31da](https://github.com/cloud-hypervisor/cloud-hypervisor/commit/4bb31daa18e5d7181c7d4e1388e101b5536db463)
> 
>     * Include missing Cargo.lock changes

Thanks @sameo  and @liuw  for helping all the way. The PR looks good to me. I have listed all the future TODO that I am planning to finish gradually.

**Comment by rbradford** [comment]
Can you open a GitHub issue with all the things that have been postponed to a later PR?

File: hypervisor/src/mshv/x86_64/mod.rs [hunk 18]
```diff
@@ -0,0 +1,154 @@
...
+        msr!(msr_index::MSR_SYSCALL_MASK),
+        msr!(msr_index::MSR_IA32_TSC),
+        // TODO: feature guard kvm and move this out of kvm
+        /*
+        msr_data!(
+            msr_index::MSR_IA32_MISC_ENABLE,
+            msr_index::MSR_IA32_MISC_ENABLE_FAST_STRING as u64
+        ),
+        msr_data!(msr_index::MSR_MTRRdefType, MTRR_ENABLE | MTRR_MEM_TYPE_WB), */
+    ])
```
Comment (rbradford) [inline hypervisor/src/mshv/x86_64/mod.rs:153]: Commented out code. Why is this here?

File: hypervisor/src/mshv/x86_64/mod.rs (hunk 2 above)
Comment (rbradford) [reply hypervisor/src/mshv/x86_64/mod.rs:146]: Did you miss acting on this feedback?

File: hypervisor/src/mshv/x86_64/mod.rs [hunk 19]
```diff
@@ -0,0 +1,154 @@
...
+pub use {
+    mshv_bindings::mshv_user_mem_region as MemoryRegion, mshv_bindings::msr_entry as MsrEntry,
+    mshv_bindings::CpuId, mshv_bindings::DebugRegisters,
+    mshv_bindings::FloatingPointUnit as FpuState, mshv_bindings::LapicState,
+    mshv_bindings::MsrList, mshv_bindings::Msrs as MsrEntries, mshv_bindings::Msrs,
+    mshv_bindings::SegmentRegister, mshv_bindings::SpecialRegisters,
+    mshv_bindings::StandardRegisters, mshv_bindings::VcpuEvents, mshv_bindings::XSave as Xsave,
+    mshv_bindings::Xcrs as ExtendedControlRegisters,
+};
+#[derive(Clone, Serialize, Deserialize)]
```
Comment (rbradford) [inline hypervisor/src/mshv/x86_64/mod.rs:25]: I think there should be a line of whitespace between the use/pub use and also between the pub use and the struct definitions.

File: hypervisor/src/cpu.rs [hunk 20]
```diff
@@ -12,13 +12,18 @@
 use crate::aarch64::VcpuInit;
 #[cfg(target_arch = "aarch64")]
 use crate::aarch64::{RegList, Register, StandardRegisters};
-use crate::{CpuState, MpState};
+#[cfg(target_arch = "x86_64")]
+use crate::x86_64::{CpuId, LapicState};
+use crate::CpuState;
+#[cfg(target_arch = "x86_64")]
+use crate::Xsave;
```
Comment (rbradford) [inline hypervisor/src/cpu.rs:20]: This whitespace line breaks the ordering.

File: hypervisor/Cargo.toml [hunk 21]
```diff
@@ -26,6 +26,8 @@ serde_json = ">=1.0.9"
 vm-memory = { version = "0.4.0", features = ["backend-mmap", "backend-atomic"] }
 vmm-sys-util = { version = ">=0.5.0", features = ["with-serde"] }
 
+epoll = ">=4.0.1"
```
Comment (rbradford) [inline hypervisor/Cargo.toml:29]: This should be in alphabetical order without the whitespace before it

File: hypervisor/src/mshv/mod.rs [hunk 22]
```diff
@@ -17,7 +17,7 @@ use crate::cpu::Vcpu;
 use crate::hypervisor;
 use crate::vm::{self, VmmOps};
 pub use mshv_bindings::*;
-use mshv_ioctls::{set_registers_64, Mshv, VcpuFd, VmFd};
+use mshv_ioctls::{set_registers_64, InterruptRequest, Mshv, VcpuFd, VmFd};
```
Comment (rbradford) [inline hypervisor/src/mshv/mod.rs:21]: Whitespace breaks the rustfmt lexical ordering

File: hypervisor/src/mshv/mod.rs [hunk 23]
```diff
@@ -35,7 +35,6 @@ pub use x86_64::*;
```
Comment (rbradford) [inline hypervisor/src/mshv/mod.rs:35]: Ditto.

File: hypervisor/src/mshv/mod.rs [hunk 24]
```diff
@@ -35,7 +35,6 @@ pub use x86_64::*;
 
 use crate::device;
```
Comment (rbradford) [inline hypervisor/src/mshv/mod.rs:37]: Ditto

File: hypervisor/src/mshv/x86_64/mod.rs (hunk 18 above)
Comment (liuw) [reply hypervisor/src/mshv/x86_64/mod.rs:147]: This should've been dropped.

**Review by liuw — APPROVED** [review]
This series LGTM.

Rob made a few cosmetic comments and I also found a typo in a commit message. To save another round-trip I've addressed Rob's comments and fixed the typo, and pushed to Muminul's branch.

**Comment by liuw** [comment]
@russell-islam Please create an issue to track future PRs like Rob suggested.

**Comment by russell-islam** [comment]
Created an issue to track followup PRs. https://github.com/cloud-hypervisor/cloud-hypervisor/issues/2048
